[JSC] Remove operationCallArityCheck calls

Constellation opened the pull request:
#### 69fd8b516f340e71e0e43b2a3f899010514bd8a6
<pre>
[JSC] Remove operationCallArityCheck calls
<a href="https://bugs.webkit.org/show_bug.cgi?id=244293">https://bugs.webkit.org/show_bug.cgi?id=244293</a>

Reviewed by Saam Barati.

We found that arity padding computation is done via runtime function and it is doing a lot of costly things (looking into CodeBlock from callee).
But since it is called from JIT, we can trivially inline it into JIT code. Plus, codeBlock-&gt;numParameters() etc. can be embedded
constant in JIT code.

It improves JetStream2 by 0.4%.

* Source/JavaScriptCore/dfg/DFGJITCompiler.cpp:
(JSC::DFG::JITCompiler::compile):
(JSC::DFG::JITCompiler::compileFunction):
* Source/JavaScriptCore/dfg/DFGJITCompiler.h:
(JSC::DFG::JITCompiler::exceptionJumpWithCallFrameRollback):
(JSC::DFG::JITCompiler::exceptionCheckWithCallFrameRollback): Deleted.
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.h:
(JSC::DFG::SpeculativeJIT::callThrowOperationWithCallFrameRollback):
(JSC::DFG::SpeculativeJIT::callOperationWithCallFrameRollbackOnException): Deleted.
(JSC::DFG::SpeculativeJIT::appendCallWithCallFrameRollbackOnException): Deleted.
(JSC::DFG::SpeculativeJIT::appendCallWithCallFrameRollbackOnExceptionSetResult): Deleted.
* Source/JavaScriptCore/ftl/FTLLink.cpp:
(JSC::FTL::link):
* Source/JavaScriptCore/jit/JIT.cpp:
(JSC::JIT::compileAndLinkWithoutFinalizing):
* Source/JavaScriptCore/jit/JIT.h:
* Source/JavaScriptCore/jit/JITInlines.h:
(JSC::JIT::appendCallWithCallFrameRollbackOnException): Deleted.

Canonical link: <a href="https://commits.webkit.org/253869@main">https://commits.webkit.org/253869@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4717aa2c4f307a9b769454ecfc410bb20d3fc926

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/87300 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/31388 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/18112 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/96425 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/149874 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/29751 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/25960 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/79413 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/91300 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/92917 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/24036 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/74095 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/23885 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/79009 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/79233 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/66901 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/79152 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/27468 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/13069 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/72833 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/27420 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/14084 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/25942 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2724 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/29106 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/36950 "Found 1 new test failure: http/wpt/crypto/unwrap-rsa-key-crash.any.worker.html") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/75653 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/29038 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/33361 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/16757 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

saambarati commented on Source/JavaScriptCore/jit/AssemblyHelpers.cpp:
```diff
@@ -1437,6 +1437,35 @@ void AssemblyHelpers::emitSaveOrCopyLLIntBaselineCalleeSavesFor(CodeBlock* codeB
+    //
+    //  Then, we return align2(padding).
+    static_assert(stackAlignmentRegisters() == 2);
+    if (WTF::roundUpToMultipleOf(stackAlignmentRegisters(), numberOfParameters + CallFrame::headerSizeInRegisters) == numberOfParameters + CallFrame::headerSizeInRegisters)
+        move(TrustedImm32(numberOfParameters), paddingOutputGPR);
+    else
+        move(TrustedImm32(numberOfParameters + 1), paddingOutputGPR);
+    sub32(paddingOutputGPR, argumentCountIncludingThisGPR, paddingOutputGPR);
```
Is there a version of sub32 that takes the TrustedImm32 as the first param? That way we may not need an unneeded move.

Constellation replied:
We have `reg - imm`, but we don't have `imm - reg`, so we need to put `imm` to reg once.

saambarati reviewed (approved):
r=me

saambarati commented on Source/JavaScriptCore/jit/AssemblyHelpers.cpp:
```diff
@@ -1437,6 +1437,35 @@ void AssemblyHelpers::emitSaveOrCopyLLIntBaselineCalleeSavesFor(CodeBlock* codeB
     spooler.finalizeGPR();
 }
 
+void AssemblyHelpers::checkArity(VM& vm, unsigned numberOfParameters, GPRReg argumentCountIncludingThisGPR, GPRReg paddingOutputGPR, GPRReg scratchGPR0, GPRReg scratchGPR1, JumpList& stackOverflow)
```
I wonder if we should name this something like "getArityPadding"

saambarati replied:
my reasoning is this, we've already checked arity at this point. And the code is assuming that (not checking for overflow in sub)

Constellation replied:
Changed.

Constellation committed:
[JSC] Remove operationCallArityCheck calls
https://bugs.webkit.org/show_bug.cgi?id=244293

Reviewed by Saam Barati.

We found that arity padding computation is done via runtime function and it is doing a lot of costly things (looking into CodeBlock from callee).
But since it is called from JIT, we can trivially inline it into JIT code. Plus, codeBlock->numParameters() etc. can be embedded
constant in JIT code.

It improves JetStream2 by 0.4%.

* Source/JavaScriptCore/dfg/DFGJITCompiler.cpp:
(JSC::DFG::JITCompiler::compile):
(JSC::DFG::JITCompiler::compileFunction):
* Source/JavaScriptCore/dfg/DFGJITCompiler.h:
(JSC::DFG::JITCompiler::exceptionJumpWithCallFrameRollback):
(JSC::DFG::JITCompiler::exceptionCheckWithCallFrameRollback): Deleted.
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.h:
(JSC::DFG::SpeculativeJIT::callThrowOperationWithCallFrameRollback):
(JSC::DFG::SpeculativeJIT::callOperationWithCallFrameRollbackOnException): Deleted.
(JSC::DFG::SpeculativeJIT::appendCallWithCallFrameRollbackOnException): Deleted.
(JSC::DFG::SpeculativeJIT::appendCallWithCallFrameRollbackOnExceptionSetResult): Deleted.
* Source/JavaScriptCore/ftl/FTLLink.cpp:
(JSC::FTL::link):
* Source/JavaScriptCore/jit/JIT.cpp:
(JSC::JIT::compileAndLinkWithoutFinalizing):
* Source/JavaScriptCore/jit/JIT.h:
* Source/JavaScriptCore/jit/JITInlines.h:
(JSC::JIT::appendCallWithCallFrameRollbackOnException): Deleted.

Canonical link: https://commits.webkit.org/253869@main
